File: nvpl_blas/f77/drot.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DROT 
!
!******************************************************************************
      program   DROT_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic        sqrt, abs

      integer          n, incx, incy, i
      parameter        (n=5, incx=1, incy=1)
      double precision s, c
      parameter        (s=0.5, c=sqrt(1-s*s))
      integer          xsize, ysize
      parameter        (xsize=1+(n-1)*abs(incx),
     &                  ysize=1+(n-1)*abs(incy))
      double precision            x(xsize), y(ysize)
      data             (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/,
     &                 (y(i), i=1,ysize)/ysize*2.0/

*       External Subroutines
      external         print_dvector 
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, incy, c, s
      call print_dvector(x,n,incx,'X')
      call print_dvector(y,n,incy,'Y')
*
*      Call DROT subroutine
      call DROT(n,x,incx,y,incy,c,s)
*
      print*
      print 101
      call print_dvector(x,n,incx,'X')
      call print_dvector(y,n,incy,'Y')

      stop
 99   format('Example: DROT for applying Givens rotation matrix')
 100  format('#### args: n=',i1,', incx=',i1,', incy=',i1,', '
     &       'c=',f5.3,', s='f5.3)
 101  format('The result of DROT: ')
      end

File: nvpl_blas/c/drot.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_drot
 *
 ******************************************************************************/
#include "example_helper.h"
#include <math.h> // sqrt

int main() {
    nvpl_int_t N = 5;
    double * X;
    double * Y;
    double s = 0.5f;
    double c = sqrt(1 - s * s);
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    printf("\nExample: cblas_drot for applying Givens rotation matrix\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", incy=%" PRId64 ", c=%f, s=%f\n", (int64_t)N,
            (int64_t)incX, (int64_t)incY, c, s);

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (N - 1) * labs(incY);
    // allocate memory
    X = (double *)malloc(len_x * sizeof(double));
    Y = (double *)malloc(len_y * sizeof(double));

    // fill data
    fill_dvector(X, N, incX);
    fill_dvector(Y, N, incY);

    // print input data
    print_dvector(X, N, incX, "X");
    print_dvector(Y, N, incY, "Y");

    // call cblas_drot
    cblas_drot(N, X, incX, Y, incY, c, s);

    // print result
    printf("\nThe result of cblas_drot:\n");
    print_dvector(X, N, incX, "X");
    print_dvector(Y, N, incY, "Y");

    // release memory
    free(X);
    free(Y);
    return 0;
}
